compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -751,10 +751,18 @@ impl<'a> Parser<'a> {
                     tail.could_be_bare_literal = true;
                     if maybe_struct_name.is_ident() && can_be_struct_literal {
                         // Account for `if Example { a: one(), }.is_pos() {}`.
+                        // expand `before` so that we take care of module path such as:
+                        // `foo::Bar { ... } `
+                        // we expect to suggest `(foo::Bar { ... })` instead of `foo::(Bar { ... })`
+                        let sm = self.sess.source_map();
+                        let before = maybe_struct_name.span.shrink_to_lo();
+                        let extend_before = sm.span_extend_prev_while(before, |t| {
+                            t.is_alphanumeric() || t == ':' || t == '_'
+                        });
                         Err(self.sess.create_err(StructLiteralNeedingParens {
                             span: maybe_struct_name.span.to(expr.span),
                             sugg: StructLiteralNeedingParensSugg {
-                                before: maybe_struct_name.span.shrink_to_lo(),
+                                before: extend_before.unwrap().shrink_to_lo(),
```

Comment:
Unwrap is probably not a good idea, will cause an ICE probably 😰 -- we shouldn't be giving a suggestion if `span_extend_prev_while` fails.

Maybe just return `None`?

Reply:
fixed :)